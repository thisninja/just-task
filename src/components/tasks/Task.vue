<template>
  <div class="task-list">
    <div v-if="!newTaskForm">
      <md-button
        class="md-raised md-accent"
        @click="newTaskForm = !newTaskForm"
        >
          {{ CREATE_NEW_TASK_TEXT }}
      </md-button>
      <md-card
        class="card-wrapper"
        v-for="task in newestFirst"
        :key="task._id"
        >
        <md-card-header>
          <div class="md-title">
            <span v-if="task.completed">✔️</span>
            <span :class="{ completed: task.completed }">
              {{ task.text }}
            </span>
          </div>
          <div
            class="md-body-2"
            v-if="task.dueDate"
            >
              {{ DUE_DATE_TEXT }}: {{ getFormattedDate(task.dueDate) }}
          </div>
          <div
            class="md-body-2"
            v-if="task.completed"
            >
              {{ COMPLETED_AT_TEXT }}: {{ getFormattedDate(task.completedAt) }}
          </div>
        </md-card-header>

        <md-card-actions>
          <md-button
            class="md-primary"
            @click="onComplete(task)">
              {{ task.completed
                ? MARK_AS_UNCOMPLETED
                : MARK_AS_COMPLETED
              }}
          </md-button>
          <md-button
            v-if="!task.completed"
            @click="onEdit(task)"
            >
              {{ EDIT_BTN_TEXT }}
          </md-button>
          <md-button
            class="md-accent"
            @click="onDelete(task)"
            >
              {{ DELETE_BTN_TEXT }}
            </md-button>
        </md-card-actions>
      </md-card>
    </div>
  </div>
</template>

<script>
import {
  mapGetters,
  mapActions,
} from 'vuex';
import {
  COMPLETED_AT_TEXT,
  CREATE_NEW_TASK_TEXT,
  MARK_AS_COMPLETED,
  MARK_AS_UNCOMPLETED,
  DUE_DATE_TEXT,
  EDIT_BTN_TEXT,
  DELETE_BTN_TEXT,
} from './constants';

export default {
  name: 'Task',
  data() {
    return {
      newTaskForm: false,
      COMPLETED_AT_TEXT,
      CREATE_NEW_TASK_TEXT,
      MARK_AS_COMPLETED,
      MARK_AS_UNCOMPLETED,
      DUE_DATE_TEXT,
      EDIT_BTN_TEXT,
      DELETE_BTN_TEXT,
    };
  },
  created () {
    this.getTasks();
  },
  computed: {
    ...mapGetters([
      'tasks',
    ]),
    newestFirst() {
      return this.tasks.slice().reverse();
    },
  },
  methods: {
    ...mapActions([
      'getTasks',
      'updateTask',
      'deleteTaskById',
    ]),
    getFormattedDate(date) {
      return new Date(date).toDateString();
    },
    onDelete({ _id }) {
      this.deleteTaskById(_id);
    },
    onComplete({ _id, text, completed, dueDate }) {
      this.updateTask(
        {
          _id,
          text,
          completed: !completed,
          dueDate
        }
      );
    },
    onEdit({ _id, text }) {
      this.selectedId = _id;
      this.selectedText = text;
      this.editTaskForm = !this.editTaskForm;
    },
  }
};
</script>

<style lang="scss" scoped>
.task-list {
  min-width: 350px;
  max-width: 700px;

  .md-card.card-wrapper {
    margin: 10px;
  }

  .completed {
    text-decoration-line: line-through;
  }
}
</style>
