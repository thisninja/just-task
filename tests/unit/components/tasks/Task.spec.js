import Vuex from 'vuex';
import VueMaterial from 'vue-material';
import { mount, createLocalVue } from '@vue/test-utils';
import Task from '@/components/tasks/Task.vue';
import dataMock from './__mocks__/dataMock';

const localVue = createLocalVue();

localVue.use(VueMaterial);
localVue.use(Vuex);

describe('Task.vue', () => {
  let wrapper;
  let store;
  let actions;
  let mutations;
  let getters;

  beforeEach(() => {
    getters = getters || {
      tasks: () => [],
    };

    actions = actions || {
      getTasks: jest.fn(),
      deleteTaskById: jest.fn(),
      updateTask: jest.fn(),
    };

    store = new Vuex.Store({
      getters,
      actions,
    });

    wrapper = mount(Task, {
      store,
      localVue,
    });
  });
  describe('snapshots', () => {
    it('should match snapshot if !editTaskForm && !newTaskForm', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should match snapshot if editTaskForm || newTaskForm', () => {
      const randomInRangeOfTwo = Math.round(Math.random());
      const rangeValues = ['editTaskForm', 'newTaskForm'];
      wrapper.vm[rangeValues[randomInRangeOfTwo]] = true;

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('methods', () => {
    describe('toggleNewTaskForm', () => {
      it('should invert newTaskForm value', () => {
        expect(wrapper.vm.newTaskForm).toBeFalsy();
        wrapper.find('.fn-create-new-task').trigger('click');

        expect(wrapper.vm.newTaskForm).toBeTruthy();
      });
    });

    describe('getFormattedDate', () => {
      it('should return formatted date', () => {
        const date = new Date().getTime();
        const formatted = new Date(date).toDateString();
        expect(wrapper.vm.getFormattedDate(date)).toEqual(formatted);
      });
    });

    describe('action handlers', () => {
      beforeAll(() => {
        getters = {
          tasks: () => dataMock.tasks
        };
      });

      describe('onComplete', () => {
        it('should call deleteTaskById action', () => {
          wrapper.find('.fn-complete-task').trigger('click');
          expect(actions.updateTask).toHaveBeenCalled();
        });

        it('should match snapshot', () => {
          expect(wrapper.html()).toMatchSnapshot();
        });
      });

      describe('onEdit', () => {
        it('should call onEdit action', () => {
          wrapper.vm.onEdit = jest.fn();
          wrapper.find('.fn-edit-task').trigger('click');
          expect(wrapper.vm.onEdit).toHaveBeenCalledWith(wrapper.vm.tasks[0]);
        });

        it('should match snapshot', () => {
          expect(wrapper.html()).toMatchSnapshot();
        });
      });

      describe('onDelete', () => {
        it('should match snapshot', () => {
          expect(wrapper.html()).toMatchSnapshot();
        });

        it('should call deleteTaskById action', () => {
          wrapper.find('.fn-delete-task').trigger('click');
          expect(actions.deleteTaskById).toHaveBeenCalled();
        });
      });
    });
  });
});
