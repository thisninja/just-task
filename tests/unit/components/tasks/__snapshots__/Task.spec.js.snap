// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task.vue methods action handlers onComplete should match snapshot 1`] = `
<div class="task-list">
  <div><button type="button" class="md-button md-raised md-accent fn-create-new-task md-theme-default">
      <div class="md-ripple">
        <div class="md-button-content">
          Create new task
        </div>
      </div>
    </button>
    <div class="md-card card-wrapper md-theme-default">
      <div class="md-card-header">
        <div class="md-title"><span>✔️</span> <span class="completed">
            Do someting 2nd time
          </span></div>
        <div class="md-body-2">
          Due date: Tue Oct 29 2019
        </div>
        <div class="md-body-2">
          Completed at: Mon Oct 07 2019
        </div>
      </div>
      <div class="md-card-actions md-alignment-right"><button type="button" class="md-button md-primary fn-complete-task md-theme-default">
          <div class="md-ripple">
            <div class="md-button-content">
              Mark as uncompleted
            </div>
          </div>
        </button>
        <!----> <button type="button" class="md-button md-accent fn-delete-task md-theme-default">
          <div class="md-ripple">
            <div class="md-button-content">
              Delete
            </div>
          </div>
        </button></div>
    </div>
    <div class="md-card card-wrapper md-theme-default">
      <div class="md-card-header">
        <div class="md-title">
          <!----> <span class="">
            Do someting 1st time
          </span></div>
        <div class="md-body-2">
          Due date: Sun Oct 27 2019
        </div>
        <!---->
      </div>
      <div class="md-card-actions md-alignment-right"><button type="button" class="md-button md-primary fn-complete-task md-theme-default">
          <div class="md-ripple">
            <div class="md-button-content">
              Mark as completed
            </div>
          </div>
        </button> <button type="button" class="md-button fn-edit-task md-theme-default">
          <div class="md-ripple">
            <div class="md-button-content">
              Edit
            </div>
          </div>
        </button> <button type="button" class="md-button md-accent fn-delete-task md-theme-default">
          <div class="md-ripple">
            <div class="md-button-content">
              Delete
            </div>
          </div>
        </button></div>
    </div>
  </div>
  <!---->
</div>
`;

exports[`Task.vue methods action handlers onDelete should match snapshot 1`] = `
<div class="task-list">
  <div><button type="button" class="md-button md-raised md-accent fn-create-new-task md-theme-default">
      <div class="md-ripple">
        <div class="md-button-content">
          Create new task
        </div>
      </div>
    </button>
    <div class="md-card card-wrapper md-theme-default">
      <div class="md-card-header">
        <div class="md-title"><span>✔️</span> <span class="completed">
            Do someting 2nd time
          </span></div>
        <div class="md-body-2">
          Due date: Tue Oct 29 2019
        </div>
        <div class="md-body-2">
          Completed at: Mon Oct 07 2019
        </div>
      </div>
      <div class="md-card-actions md-alignment-right"><button type="button" class="md-button md-primary fn-complete-task md-theme-default">
          <div class="md-ripple">
            <div class="md-button-content">
              Mark as uncompleted
            </div>
          </div>
        </button>
        <!----> <button type="button" class="md-button md-accent fn-delete-task md-theme-default">
          <div class="md-ripple">
            <div class="md-button-content">
              Delete
            </div>
          </div>
        </button></div>
    </div>
    <div class="md-card card-wrapper md-theme-default">
      <div class="md-card-header">
        <div class="md-title">
          <!----> <span class="">
            Do someting 1st time
          </span></div>
        <div class="md-body-2">
          Due date: Sun Oct 27 2019
        </div>
        <!---->
      </div>
      <div class="md-card-actions md-alignment-right"><button type="button" class="md-button md-primary fn-complete-task md-theme-default">
          <div class="md-ripple">
            <div class="md-button-content">
              Mark as completed
            </div>
          </div>
        </button> <button type="button" class="md-button fn-edit-task md-theme-default">
          <div class="md-ripple">
            <div class="md-button-content">
              Edit
            </div>
          </div>
        </button> <button type="button" class="md-button md-accent fn-delete-task md-theme-default">
          <div class="md-ripple">
            <div class="md-button-content">
              Delete
            </div>
          </div>
        </button></div>
    </div>
  </div>
  <!---->
</div>
`;

exports[`Task.vue methods action handlers onEdit should match snapshot 1`] = `
<div class="task-list">
  <div><button type="button" class="md-button md-raised md-accent fn-create-new-task md-theme-default">
      <div class="md-ripple">
        <div class="md-button-content">
          Create new task
        </div>
      </div>
    </button>
    <div class="md-card card-wrapper md-theme-default">
      <div class="md-card-header">
        <div class="md-title"><span>✔️</span> <span class="completed">
            Do someting 2nd time
          </span></div>
        <div class="md-body-2">
          Due date: Tue Oct 29 2019
        </div>
        <div class="md-body-2">
          Completed at: Mon Oct 07 2019
        </div>
      </div>
      <div class="md-card-actions md-alignment-right"><button type="button" class="md-button md-primary fn-complete-task md-theme-default">
          <div class="md-ripple">
            <div class="md-button-content">
              Mark as uncompleted
            </div>
          </div>
        </button>
        <!----> <button type="button" class="md-button md-accent fn-delete-task md-theme-default">
          <div class="md-ripple">
            <div class="md-button-content">
              Delete
            </div>
          </div>
        </button></div>
    </div>
    <div class="md-card card-wrapper md-theme-default">
      <div class="md-card-header">
        <div class="md-title">
          <!----> <span class="">
            Do someting 1st time
          </span></div>
        <div class="md-body-2">
          Due date: Sun Oct 27 2019
        </div>
        <!---->
      </div>
      <div class="md-card-actions md-alignment-right"><button type="button" class="md-button md-primary fn-complete-task md-theme-default">
          <div class="md-ripple">
            <div class="md-button-content">
              Mark as completed
            </div>
          </div>
        </button> <button type="button" class="md-button fn-edit-task md-theme-default">
          <div class="md-ripple">
            <div class="md-button-content">
              Edit
            </div>
          </div>
        </button> <button type="button" class="md-button md-accent fn-delete-task md-theme-default">
          <div class="md-ripple">
            <div class="md-button-content">
              Delete
            </div>
          </div>
        </button></div>
    </div>
  </div>
  <!---->
</div>
`;

exports[`Task.vue snapshots should match snapshot if !editTaskForm && !newTaskForm 1`] = `
<div class="task-list">
  <div><button type="button" class="md-button md-raised md-accent fn-create-new-task md-theme-default">
      <div class="md-ripple">
        <div class="md-button-content">
          Create new task
        </div>
      </div>
    </button> </div>
  <!---->
</div>
`;

exports[`Task.vue snapshots should match snapshot if editTaskForm || newTaskForm 1`] = `
<div class="task-list">
  <!---->
  <div>
    <div class="card-body">
      <div class="card-text"><span>
      Add task
    </span>
        <div class="md-field md-theme-default md-clearable md-datepicker"><i class="md-icon md-icon-font md-icon-image md-date-icon md-theme-default"><svg height="24" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg">
              <path d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"></path>
              <path d="M0 0h24v24H0z" fill="none"></path>
            </svg></i> <input type="text" pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}" id="md-input-sw8m4l7pw" class="md-input"> <label for="md-input-sw8m4l7pw">
            Due date
          </label>
          <!---->
          <!---->
          <!---->
          <!---->
        </div>
        <div class="md-field md-theme-default"><label for="md-input-953nbnd89">
            Enter your task description
          </label> <input type="text" id="md-input-953nbnd89" class="md-input">
          <!---->
          <!---->
          <!---->
        </div> <button disabled="disabled" type="button" class="md-button fn-save md-theme-default">
          <div class="md-ripple md-disabled">
            <div class="md-button-content">
              Save
            </div>
          </div>
        </button> <button type="button" class="md-button fn-cancel md-theme-default">
          <div class="md-ripple">
            <div class="md-button-content">
              Cancel
            </div>
          </div>
        </button></div>
    </div>
  </div>
</div>
`;
